src/rpc/message.cpp

```diff
@@ -211,82 +140,89 @@ cryptonote::rpc::error FullMessage::getError()
   return err;
 }
 
-FullMessage FullMessage::requestMessage(const std::string& request, Message* message)
+std::string FullMessage::getRequest(const std::string& request, const Message& message, const unsigned id)
 {
-  return FullMessage(request, message);
-}
+  rapidjson::StringBuffer buffer;
+  {
+    rapidjson::Writer<rapidjson::StringBuffer> dest{buffer};
 
-FullMessage FullMessage::requestMessage(const std::string& request, Message* message, rapidjson::Value& id)
-{
-  auto mes = requestMessage(request, message);
-  mes.setID(id);
-  return mes;
-}
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, jsonrpc, std::string{"2.0"});
 
-FullMessage FullMessage::responseMessage(Message* message)
-{
-  return FullMessage(message);
-}
+    dest.Key(id_field);
+    json::toJsonValue(dest, id);
 
-FullMessage FullMessage::responseMessage(Message* message, rapidjson::Value& id)
-{
-  auto mes = responseMessage(message);
-  mes.setID(id);
-  return mes;
+    dest.Key(method_field);
+    json::toJsonValue(dest, request);
+
+    dest.Key(params_field);
+    message.toJson(dest);
+
+    dest.EndObject();
+
+    if (!dest.IsComplete())
+      throw std::logic_error{"Invalid JSON tree generated"};
+  }
+  return std::string{buffer.GetString(), buffer.GetSize()};
 }
 
-FullMessage* FullMessage::timeoutMessage()
+
+std::string FullMessage::getResponse(const Message& message, const rapidjson::Value& id)
 {
-  auto *full_message = new FullMessage();
+  rapidjson::StringBuffer buffer;
+  {
```

Comment:
Any specific reason for this added scope? Looking at it I started to assume that the `rapidjson::Writer` destructor performs relevant tasks (e.g. writing to the buffer), but I couldn't quickly find anything related to that.

Reply:
In case the `writer<..>` needs to flush anything on destruction. It should be unnecessary in this situation.